rfd/0038-parquet-and-events.md

```diff
@@ -0,0 +1,139 @@
+
+#### Key management
+
+Since customers may have tools that are designed to work specifically with different cloud services and their encryption and key management schemes we may need to slightly tailor the key management approach based on the cloud service the deployment is connected to. Described below is the default key management implementation that works with all supported storage backends.
+
+When using symmetric encryption algorithms to encrypt and decrypt data on auth servers a strategy is needed for managing these keys and making sure all auth servers in a high availability deployment can decrypt data from all other auth servers.
+
+Broadly, there are two options available:
+
+1. Sharing a single encryption key across all authentication servers.
+2. Taking a card from the AWS Encryption SDK playbook and using a two layer key system.
+
+Option 1 has multiple significant drawbacks which make it unsuitable for production use. The major issue is that all data is fundamentally tied to the only layer of security meaning that if a malicous party potentially acquired access hold to the key, all data would have to be reincrypted. This makes key rotation prohibitive. Another major issue is that of prolonged key reuse which contradicts industry best practices.
+
+Option 2 is doing something very similar to what the AWS Encryption SDK does. Ideally using that library would be ideal since it is a very nice encryption toolkit but unfortunately it isn't available for Go. The strategy involves using unique ephemeral data keys for each encrypted object (rotated when reincrypted/modified) that are encrypted and bundled with each encrypted object in a defined format. A central PEM encoded master keypair stored in a new resource is used to encrypt these per object keys.
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.key` stored next to the encrypted object.
+
+Each encrypted object has an accompanying file named `$objectName.key` that stores a JSON document of this format.
+
+```json
+{
+    // Contains a base64 encoded AES-CBC-HMACSHA256 data key encrypted with the master key.
+    "dataKey": "string",
+
+    // An integer starting at 1 one that communicates the schema used for encryption.
+    // May be changed in the future.
+    "version": "number",
+
+    // A hex encoded version of the SHA-256 hash of the master key.
+    // This is used to determine state during key rotation.
+    "masterKey": "string",
+}
+```
```

Comment:
In the interest of smooth rotation, how about supporting multiple "recipients" (i.e. multiple `dataKey`s encrypted for different public keys).

That would allow us to perform rotations entirely asynchronously, because we could add a recipient, wait until all files have been updated with the new recipient (during which time new files would be encrypted for both recipients), and then purge the old recipient afterwards.  No weird intermediate state if we crash in the middle of a migration, or one auth server is lagging behind the others.